proof-systems/src/gm17/examples/recursive-snark/gm17.rs

```diff
@@ -0,0 +1,321 @@
+    type PairingGadgetTick = MNT6753PairingGadget;
+    type PairingGadgetTock = MNT4753PairingGadget;
+    const TICK_CURVE: &'static str = "MNT6_753";
+    const TOCK_CURVE: &'static str = "MNT4_753";
+}
+
+fn main() -> Result<(), Box<dyn Error>> {
+    let args: Vec<String> = env::args().collect();
+    if args.len() < 3 || args[1] == "-h" || args[1] == "--help" {
+        println!(
+            "\nHelp: Invoke this as <program> <num_constraints> <output_file_path> [<order>]\n"
+        );
+        println!("<order> defines the order in which the MNT4/6 curves should be used:");
+        println!("46 (default) uses the MNT4_753 curve for the inner and outer circuit;");
+        println!("64 uses the MNT6_753 curve for the inner and outer circuit.");
+        return Ok(());
+    }
+    let num_constraints: usize = args[1].parse().unwrap();
+    let output_file_path = PathBuf::from(args[2].clone());
+
+    if args.len() < 4 || args[3] == "46" {
+        run::<MNT46>(num_constraints, output_file_path)
+    } else {
+        run::<MNT64>(num_constraints, output_file_path)
+    }
+}
+
+fn run<C: CurvePair>(
+    num_constraints: usize,
+    output_file_path: PathBuf,
+) -> Result<(), Box<dyn Error>> {
+    let mut wtr = if !output_file_path.exists() {
+        println!("Creating output file");
+        let f = OpenOptions::new()
+            .create(true)
+            .append(true)
+            .open(output_file_path)?;
+        let mut wtr = csv::Writer::from_writer(f);
+        wtr.write_record(&[
+            "num_constraints",
+            "setup_inner",
+            "prover_inner",
+            "setup_middle",
+            "prover_middle",
+            "setup_outer",
+            "prover_outer",
+            "verifier_outer",
+        ])?;
+        wtr
+    } else if output_file_path.is_file() {
+        let f = OpenOptions::new().append(true).open(output_file_path)?;
+        csv::Writer::from_writer(f)
+    } else {
+        println!("Path to output file does not point to a file.");
+        process::exit(1);
+    };
+    // This may not be cryptographically safe, use
+    // `OsRng` (for example) in production software.
+    let rng = &mut thread_rng();
+
+    // Let's benchmark stuff!
+    let samples = 1;
+    let mut total_setup_inner = Duration::new(0, 0);
+    let mut total_proving_inner = Duration::new(0, 0);
+    let mut total_setup_middle = Duration::new(0, 0);
+    let mut total_proving_middle = Duration::new(0, 0);
+    let mut total_setup_outer = Duration::new(0, 0);
+    let mut total_proving_outer = Duration::new(0, 0);
+    let mut total_verifying_outer = Duration::new(0, 0);
+
+    // Just a place to put the proof data, so we can
+    // benchmark deserialization.
+    // let mut proof_vec = vec![];
+
+    for sample in 0..samples {
+        println!("Running sample {}/{}", sample + 1, samples);
+        let mut inputs: Vec<<C::PairingEngineTick as PairingEngine>::Fr> =
+            Vec::with_capacity(num_constraints);
```

Comment:
As in the Groth16 example: very atypical to bench the base circuit using num_constraints inputs.